src/python/pants/option/arg_splitter.py

```diff
@@ -168,37 +152,35 @@ def assign_flag_to_scope(flg: str, default_scope: str) -> None:
             assign_flag_to_scope(flag, GLOBAL_SCOPE)
         scope, flags = self._consume_scope()
         while scope:
-            if not self._check_for_help_request(scope.lower()):
-                add_scope(scope)
-                if scope in self._known_goal_scopes:
-                    goals.add(scope)
-                else:
-                    unknown_scopes.append(scope)
-                for flag in flags:
-                    assign_flag_to_scope(flag, scope)
+            add_goal(scope)
+            for flag in flags:
+                assign_flag_to_scope(flag, scope)
             scope, flags = self._consume_scope()
 
         while self._unconsumed_args and not self._at_double_dash():
-            arg = self._unconsumed_args.pop()
-            if arg.startswith("-"):
+            if self._at_flag():
+                arg = self._unconsumed_args.pop()
                 # We assume any args here are in global scope.
-                if not self._check_for_help_request(arg):
-                    assign_flag_to_scope(arg, GLOBAL_SCOPE)
-            elif self.likely_a_spec(arg):
+                assign_flag_to_scope(arg, GLOBAL_SCOPE)
+                continue
+
+            arg = self._unconsumed_args.pop()
+            if self.likely_a_spec(arg):
                 specs.append(arg)
-            elif arg not in self._known_scopes:
-                unknown_scopes.append(arg)
+            else:
+                add_goal(arg)
```

Comment:
This has slightly modified behaviour compared to before.. as any known goals presented after specs were silently ignored before, where as they are now acted upon. 

But I think this needs some more work, as any options for those goals after any specs will be for the global scope only, which will feel weird, if you start using it, as these two will *not* be equivalent:
`./pants run --args="foo" src/my:target`
vs
`./pants src/my:target run --args="foo"`

Reply:
Right. I like that approach. Makes it clean, and more flexible in terms of what the built-ins can do.

Reply:
> That would lose the ability to do things like `./pants help $subsystem --global-option` ... but that doesn't seem too bad.

Hm... shoot. Thinking about this more, I think that this means that builtin goals end up needing to implement their own argument parsing, which is less than ideal. Although `help` doesn't need to consume any modifiers, I had forgotten in this comment that `gc` will likely have modifier flags.

On the other hand, this is a fairly clear way for some builtin goals to have subcommands without native support for them. `gc` could _probably_ have all of its modifiers implemented via flags. Some goals using subcommands, while others don't seems like a recipe for confusion, unfortunately... so I think that I'd continue to want to avoid subcommands for now.

So... I might have been wrong here having parsing short-circuit like this. `BuiltinGoals` being able to accept arguments using our existing `Options` facilities seems pretty important.

Reply:
Makes sense. As I see it, we want a hybrid here, as the first builtin goal encountered is the one being used, but keep parsing goals/options until end/or pass through args. I'll update the PR with this.